Improving voting stats

fcamblor opened the pull request:
Hi there, and happy new year 🎉 !

Here is a PR that brings some improvements on the "MyVotes" pages (both for comitee and GTs).

First, a bug I discovered long time ago in BDX I/O fork, and which allow GT to review "special" talks (`Keynote` & `Other` proposal types) once they have finished reviewing every talks (see https://twitter.com/fcamblor/status/827258007167700992 :-))
Not sure if this was an easter egg intentionally done (I can remove the commit if it was the case, no problem : just tell me)

Second, fixed a bug on average/stddev stats :
<img width="1177" alt="All_my_votes_-_Devoxx_France_2020" src="https://user-images.githubusercontent.com/603815/71645070-cc443d80-2cd2-11ea-98ae-a2015c8c58d9.png">

Third, I moved the proposal type selection above the stats
<img width="975" alt="All_my_votes_-_Devoxx_France_2020" src="https://user-images.githubusercontent.com/603815/71645428-ad48aa00-2cd8-11ea-8f95-4b5c599e6aae.png">
At the same time, I removed from the list the proposal types that are not supposed to be accessible to GTs.

Last, but not least, I improved some stats :
- Showing number of remaining talks to review (because I like to see the progression of my reviews) ...
- ... with a button allowing to review one of them when there is at least one of them. Typically, I finished reviewing every TiA, but new proposals are created everyday ... and the only option to keep my remaining TiA to review to 0 is to look for new TiA on the proposals list which contains currently a lot of Conferences. I will use this button as a shortcut to quickly find some proposals to review in a specific format :-)


Important notes :
- Like previous PRs, I was not able to test it locally as I don't have any Redis DB with some meaningful data inside
- On this particular PR, I was even not able to verify if everything was compiling ... for an unknown reason, my workspace is completely broken under intellij and I get a lot of errors while trying to compile everything (not sure if the SBT / Play2 integration is still ok)
- Don't hesitate to review commits by commits : I tried to keep them small and tell a "story" to help the review :-)
- There is one "functional regression" that is introduced in f221f10e84f08d8a41b8739a2b11cae1840c0b12 : prior to this commit, I assume that if you voted on a talk which was, later, deleted, then you were able to still see your vote on the "My votes". After this commit, this is no longer the case (for simplification, I evicted reviewed proposals which were in "invalid" status). Tell me if this was made on purpose, I can revert this change easily.

fcamblor commented on app/models/ReviewByGoldenTicket.scala:
```diff
@@ -104,15 +104,12 @@ object ReviewByGoldenTicket {
-      ReviewByGoldenTicket.allProposalsNotReviewed(reviewerUUID)
-        .filter(p => p.talkType == ConferenceDescriptor.ConferenceProposalTypes.CONF || p.talkType == ConferenceDescriptor.ConferenceProposalTypes.TIA || p.talkType == ConferenceDescriptor.ConferenceProposalTypes.LAB || p.talkType == ConferenceDescriptor.ConferenceProposalTypes.QUICK || p.talkType == ConferenceDescriptor.ConferenceProposalTypes.UNI)
-        .filterNot(_.sponsorTalk)
-    }
```
@nicmarti I'm not 100% of the assumption I made about this simplification here, I mean :

isCFPOpen=true => black listing KEY and OTHER proposal types
isCFPOpen=false => white listing CONF / TIA / LAB / QUICK / UNI proposal types
implicitely: GT have never the right to vote on KEY & OTHER proposal types, and are not able to vote on BOF either once CFP is closed (this latest rule sounds weird to me)

If that was not the idea behind the if/else, don't hesitate to tell me :-)

fcamblor commented on app/models/Review.scala:
```diff
@@ -304,23 +304,13 @@ object Review {
         proposalId: String =>
           val score = Option(client.zscore(s"Proposals:Votes:$proposalId", reviewerUUID))
           score match {
-            case None =>
```
@nicmarti I'm not 100% sure of my change here, as I moved the result of this method from a `Set[(String, Option[Double])]` to a `Set[(String, Double)]` : tldr; votes cannot be "empty" anymore.

By reading the code, I understood that we were in the "score is empty" case only when `client.zscore()` jedis call was returning an empty score.
I'm not sure if this is related to some legacy behaviour, or if, when you vote with an "abstention", you're typically in that case (in this latest case, I think it means that I introduced a regression here)

fcamblor replied:
Reworked the change with 0461e109a0d7839144c266b7ab4f68ad76bc47b6 in order to keep the "empty score => 0 score" transformation

nicmarti replied:
If I remember well, you cannot use flatMap here, as you want to count the number of elements. I suggest we keep the existing code and we don't change this .

fcamblor replied:
@nicmarti not sure to understand which `flatMap` you talk about :-)

But anyway, doest 0461e10 look better to you ? 
I think I better captured the spirit of the initial code here, I mean transforming :
```
def allVotesFromUser(reviewerUUID: String): Set[(String, Option[Double])] = Redis.pool.withClient {
  implicit client =>
    client.smembers(s"Proposals:Reviewed:ByAuthor:$reviewerUUID").flatMap {
      proposalId: String =>
        val score = Option(client.zscore(s"Proposals:Votes:$proposalId", reviewerUUID))
        score match {
          case None =>
            val state = Proposal.findProposalState(proposalId)
            state.flatMap {
              case ProposalState.DRAFT => None
              case ProposalState.DECLINED => None
              case ProposalState.DELETED => None
              case ProposalState.REJECTED => None
              case ProposalState.ARCHIVED => None
              case ProposalState.UNKNOWN => None
              case _ => Option((proposalId, None))
            }
          case Some(_) =>
            val state = Proposal.findProposalState(proposalId)
            state.flatMap {
              case ProposalState.DRAFT => None
              case ProposalState.DECLINED => None
              case ProposalState.DELETED => None
              case ProposalState.REJECTED => None
              case ProposalState.ARCHIVED => None
              case ProposalState.UNKNOWN => None
              case _ =>
                Option((proposalId, score.map(_.toDouble)))
            }
        }
    }
}
```
plus the 
```
val sortedListOfProposals = allMyVotesForSpecificProposalType.toList.sortBy {
  case (proposalID, maybeScore) =>
    maybeScore.getOrElse(0.toDouble)
}.reverse
```
transformation in `CFPAdmin` ....

... into :
```
def allVotesFromUser(reviewerUUID: String): Set[(String, Double)] = Redis.pool.withClient {
  implicit client =>
    client.smembers(s"Proposals:Reviewed:ByAuthor:$reviewerUUID").flatMap {
      proposalId: String =>
        val score = Option(client.zscore(s"Proposals:Votes:$proposalId", reviewerUUID))
        val state = Proposal.findProposalState(proposalId)
        state.flatMap {
          case ProposalState.DRAFT => None
          case ProposalState.DECLINED => None
          case ProposalState.DELETED => None
          case ProposalState.REJECTED => None
          case ProposalState.ARCHIVED => None
          case ProposalState.UNKNOWN => None
          case _ => Option((proposalId, score.map(_.toDouble).getOrElse(0)))
        }
    }
}
```
_(+ updating sort in `CFPAdmin` controller)_

nicmarti commented:
I let @aheritier have a look at it before I merge

aheritier commented on app/models/ConferenceDescriptor.scala:
```diff
@@ -49,6 +49,7 @@ case class TrackDesc(id: String, imgSrc: String, i18nTitleProp: String, i18nDesc
 
 case class ProposalConfiguration(id: String, slotsCount: Int,
                                  givesSpeakerFreeEntrance: Boolean,
+                                 accessibleTypeToGoldenTicketReviews: () => Boolean,
```
@fcamblor A compilation error is because you forgot to set a value line 63

fcamblor replied:
fixed through d66ef0181df83cd95d6d25f33659015efdbe4aeb

I also proposed to colocalize `UNKNOWN` configuration with other configs in e4c646301581e9fa24b4edea94d506a5425ad6d4 in order to avoid such omission in the future.
Problem with this impl is this is a bit verbose and we "lose" the `ProposalType` companion class.
Another way to do it is visible [here](https://github.com/fcamblor/cfp-devoxx/commit/7fc214a98d7092ffb60320d59e6fcaed017cbcc6) but it introduces a weird chicken-and-egg import [here](https://github.com/fcamblor/cfp-devoxx/commit/7fc214a98d7092ffb60320d59e6fcaed017cbcc6#diff-bdb737088b27e60c0338d67e9789d7c9R5) ... I can swap e4c646301581e9fa24b4edea94d506a5425ad6d4 with 7fc214a98d7092ffb60320d59e6fcaed017cbcc6 if you pref, just tell me

aheritier commented on app/models/Review.scala:
```diff
@@ -304,35 +304,20 @@ object Review {
-  def allVotesFromUser(reviewerUUID: String): Set[(String, Option[Double])] = Redis.pool.withClient {
+  def allVotesFromUser(reviewerUUID: String): Set[(String, Double)] = Redis.pool.withClient {
     implicit client =>
       client.smembers(s"Proposals:Reviewed:ByAuthor:$reviewerUUID").flatMap {
```
Compilation failure:

```
Error:(309, 77) type mismatch;
 found   : scala.collection.immutable.Set[(String, AnyVal)]
 required: Set[(String, Double)]
Note: (String, AnyVal) >: (String, Double), but trait Set is invariant in type A.
You may wish to investigate a wildcard type such as `_ >: (String, Double)`. (SLS 3.2.10)
      client.smembers(s"Proposals:Reviewed:ByAuthor:$reviewerUUID").flatMap {
```

aheritier replied:
The problem is in:

```
state.flatMap {
            case ProposalState.DRAFT => None
            case ProposalState.DECLINED => None
            case ProposalState.DELETED => None
            case ProposalState.REJECTED => None
            case ProposalState.ARCHIVED => None
            case ProposalState.UNKNOWN => None
            case _ => Option((proposalId, score.map(_.toDouble).getOrElse(0)))
          }
```

There is a mixup of `Option((proposalId, score.map(_.toDouble).getOrElse(0)))` and `None`
What you want is to not add these states thus you should convert the IDs to Proposals then filter them then map them ?

```

fcamblor replied:
fixed with ca55be32b2ca0c4583535a84985cf7f3bcb72901

aheritier commented on app/models/ReviewByGoldenTicket.scala:
```diff
@@ -262,35 +270,20 @@ object ReviewByGoldenTicket {
-  def allVotesFromUser(reviewerUUID: String): Set[(String, Option[Double])] = Redis.pool.withClient {
+  def allVotesFromUser(reviewerUUID: String): Set[(String, Double)] = Redis.pool.withClient {
     implicit client =>
       client.smembers(s"ReviewGT:Reviewed:ByAuthor:$reviewerUUID").flatMap {
```
Same problem here like in Review.scala :

```
Error:(275, 76) type mismatch;
 found   : scala.collection.immutable.Set[(String, AnyVal)]
 required: scala.collection.immutable.Set[(String, Double)]
Note: (String, AnyVal) >: (String, Double), but trait Set is invariant in type A.
You may wish to investigate a wildcard type such as `_ >: (String, Double)`. (SLS 3.2.10)
      client.smembers(s"ReviewGT:Reviewed:ByAuthor:$reviewerUUID").flatMap {
```

fcamblor replied:
fixed with ca55be32b2ca0c4583535a84985cf7f3bcb72901

aheritier commented on app/views/GoldenTicketController/allMyGoldenTicketVotes.scala.html:
```diff
@@ -13,21 +13,25 @@ <h1 class="panel-title"><i class="fas fa-chart-bar"></i> @Messages("gt.yourVotes
 
-                        @defining(allMyVotes.filter(_._2.isDefined).filterNot(_._2 == Some(0)).toList.map(_._2.get)){allScores=>
-                            @Messages("gt.youVoted", allMyVotes.count(s => s._2.isDefined && s._2 != Some(0)))<br>
+                        @defining(allMyVotesExcludingAbst){allScores=>
```
```suggestion
                        @defining(allMyVotesExcludingAbst.map(_._2)){allScores=>
```

It's not compiling otherwise :

```
Error:(51, 102) type mismatch;
 found   : List[(String, Double)]
 required: List[Double]
                            """),_display_(/*30.30*/Messages("gt.averageVote", library.Stats.average(allScores))),format.raw/*30.90*/("""
```

It is the change you did in the CFP side

fcamblor replied:
I really need to setup my dev environment correctly... at least to avoid such compilation failures

Regarding DB : this evening, I will try to setup again the dev environment with an empty Redis (had a look at the `README` and it seems like I should be able to start something from scratch)
Just crossing fingers about the fact that I don't need data populated into the Redis in order to make it work :-)

fcamblor replied:
OK I setup my dev environment and fixed/pushed the compilation fixes (fixed current compilation error with 19bc1deb4f7c7773bb4d2f89acd923eb8b46e229)

![All_my_votes_-_Devoxx_France_2020](https://user-images.githubusercontent.com/603815/71674840-a3788280-2d7c-11ea-982b-b20e670e9dfb.png)
![All_my_votes_-_Devoxx_France_2020](https://user-images.githubusercontent.com/603815/71674858-ad9a8100-2d7c-11ea-8dd9-3685f5f306b7.png)


I think it will be easier for me now (I was completely blind until now)

aheritier commented:
In term of code review it looks good to me @fcamblor and I like your changes.
But we have a problem, which isn't (as far as I can see introduced by your changes cc @nicmarti )
I was thinking it was a problem in your code or in my DB sync but it's not 
If I go on the home page in production I have:

<img width="1089" alt="Screenshot 2020-01-03 01 16 08" src="https://user-images.githubusercontent.com/174600/71700978-21617b80-2dc7-11ea-95b6-3b62f1e82145.png">

387 reviewed / 380 voted / 9 to review

And then in My votes:

<img width="515" alt="Screenshot 2020-01-03 01 17 31" src="https://user-images.githubusercontent.com/174600/71700982-27575c80-2dc7-11ea-81fb-5cd9d314eb86.png">

You have voted for 237 talks
Sum of all your points: 1524
Your average vote is 6.43
Standard deviation for your vote is 1.66

Also the homepage tells me that I have 9 talks to review but my reviews only 4

And to finish in the leaderbord reviewers

<img width="200" alt="Screenshot 2020-01-03 01 18 30" src="https://user-images.githubusercontent.com/174600/71700988-2de5d400-2dc7-11ea-86d0-7b9cf3ab441d.png">

380 Reviews (thus the voted ones in the homepage?). But why 237 in My votes ?
Average rate: 6.75 ... Not sure why it's 6.43 in My votes
Points: 2566 ? but 1524 in My votes ....

fcamblor commented:
Will try to have a look at your inconsistencies tomorrow, and see if I can reproduce / guess from the code what's going on :)

fcamblor commented:
Oh I think I get it... "my votes" is scoped to a specific proposal type (by default, conferences)

That's why you only see conferences stats by default. 

Having proposal type selection coming "after" the stats is really misleading, and that's.something I fixed in the UI. 

Will investigate on the 9 VS 4 talks to review tomorrow, time to sleep now ;)

aheritier commented:
Old or the number of reviews it must be the same. I effectively completely
missed that it was necessary for a type of talk. Your change is thus very
useful.

Le ven. 3 janv. 2020 à 02:07, Frédéric Camblor <notifications@github.com> a
écrit :

> Oh I think I get it... "my votes" is scoped to a specific proposal type
> (by default, conferences)
>
> That's why you only see conferences stats by default.
>
> Having proposal type selection coming "after" the stats is really
> misleading, and that's.something I fixed in the UI.
>
> Will investigate on the 9 VS 4 talks to review tomorrow, time to sleep now
> ;)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nicmarti/cfp-devoxx/pull/194?email_source=notifications&email_token=AABKUCBFLBO7XB7X7LMCVALQ32FWDA5CNFSM4KB4MGPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH77YNA#issuecomment-570424372>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABKUCABECC5IT2AYQCPUHDQ32FWDANCNFSM4KB4MGPA>
> .
>
-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

fcamblor commented:
rebased +force pushed by merging fixup commits

@aheritier when you say :
> Also the homepage tells me that I have 9 talks to review but my reviews only 4

Does it mean that in your Admin home, you have "9 talks to review" which is displayed, and only 4 proposals appear in the remaining proposal list of talks to review ?
(to be sure to understand well)

aheritier commented:
It’s good to me @fcamblor
I think we can merge it @nicmarti
Now I better understand that “my votes” page is per type of talk only.

Le ven. 3 janv. 2020 à 12:06, Frédéric Camblor <notifications@github.com> a
écrit :

> rebased +force pushed by merging fixup commits
>
> @aheritier <https://github.com/aheritier> when you say :
>
> Also the homepage tells me that I have 9 talks to review but my reviews
> only 4
> Does it mean that in your Admin home, you have "9 talks to review" which
> is displayed, and only 4 proposals appear in the remaining proposal list of
> talks to review ?
> (to be sure to understand well)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nicmarti/cfp-devoxx/pull/194?email_source=notifications&email_token=AABKUCBMPKTVQOIQJCFZXNTQ34L2DA5CNFSM4KB4MGPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIA4TFI#issuecomment-570542485>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABKUCCL5UDRBSLONE2TDIDQ34L2DANCNFSM4KB4MGPA>
> .
>
-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier

nicmarti commented:
@fcamblor Great work ! From what I understand, we need to merge both new attributes into ConferenceDescriptor. However I would prefer that you confirm, since I merged #196 before this one. Thanks for your work.

fcamblor commented:
@nicmarti yes I'm going to rebase current branch on `dev` in order to fix conflicts :-)

In the meantime, I will add 2-3 commits (I was going to push them just before you started merging everything ... just in time :-))

I'll tell you once ready (will maybe need some latest review from @aheritier before merge)

fcamblor commented:
@aheritier FYI, I pushed 3 additionnal commits to introduce sub-filtering by track on myvotes screen.

May be something useful for proposal types that have a large amount of proposals (conferences, I'm watching you :-))

<img width="1140" alt="All_my_votes_-_Devoxx_France_2020" src="https://user-images.githubusercontent.com/603815/71782789-74614b80-2fde-11ea-89f7-7b538c480789.png">

fcamblor commented:
OK I rebased over `dev` current PR

Prior to merge, let's wait for @aheritier latest tests on some addition I introduced during the rebase :
- Track sub-filtering on My Votes screen (see [this comment](https://github.com/nicmarti/cfp-devoxx/pull/194#issuecomment-570925306))
- Some small color Improvements on proposal score summary (see [this comment](https://github.com/nicmarti/cfp-devoxx/pull/197#issuecomment-570904727))

aheritier commented:
It's all good for me @fcamblor. Filtering is working fine
Well done cc @nicmarti

Not sure why on my screen the buttons are on two lines when I have an ABS but it's really not a big issue :-)

<img width="1127" alt="Screenshot 2020-01-05 19 42 58" src="https://user-images.githubusercontent.com/174600/71784553-ee043400-2ff4-11ea-8ce2-bba69d946787.png">

When @nicmarti will have merged it could send an email to the program team to explain to them these changes.
I am pretty sure that they aren't using the "my votes" today because it wasn't really powerful. With what you did it can be very useful

fcamblor commented:
@aheritier Wow I need to add some pixels to the column size (it was fitting for "10", not for "Abs")

fcamblor commented:
@aheritier done in aa99581

aheritier commented:
👍

Le dim. 5 janv. 2020 à 20:09, Frédéric Camblor <notifications@github.com> a
écrit :

> @aheritier <https://github.com/aheritier> done in aa99581
> <https://github.com/nicmarti/cfp-devoxx/commit/aa99581153a7019701fa6fe677836037e9610484>
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nicmarti/cfp-devoxx/pull/194?email_source=notifications&email_token=AABKUCGNE6R6B6HRDDB3Z3TQ4IV55A5CNFSM4KB4MGPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEID5I6A#issuecomment-570938488>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABKUCEA2WZM3OOWWJ33ULTQ4IV55ANCNFSM4KB4MGPA>
> .
>
-- 
-----
Arnaud Héritier
http://aheritier.net
Mail/GTalk: aheritier AT gmail DOT com
Twitter/Skype : aheritier
